### DISCUSSION TITLE
How share utils helpers between tasks?

### DISCUSSION THREAD

**Discussion by Ernest0-Production** [jdx/mise#3417] [Q&A]
Hi! I'm trying to figure out how I could automatically load a set of files and/or a file that contains helper functions that will be automatically loaded into all tasks, without having to explicitly call `source myutils.sh`.

Please tell me, does mise allow you to do such a trick?

**Comment by jdx** [comment]
I don't know or have many ideas, perhaps others can brainstorm or come up with recipes

**Comment by syhol** [comment]
One option is to convert those functions into individual executable shell scripts in a new directory, then add that directory to the path with https://mise.jdx.dev/environments.html#env-path any shared env vars can be directly set in the config.

E.g.
From

`utils.sh`
```bash
#!/usr/bin/env bash

export PROJECT = "foo"

function log {
  echo "[$PROJECT] [$(date +%FT%T)] $1"
}

function compose {
  docker-compose --project $PROJECT --volume $(pwd):/srv ${@}
}
```

To

`mise.toml`
```toml
[env]
PROJECT = "foo"
PROJECT_ROOT = "{{ config_root }}"
_.path = "./utils"
```

`utils/log` (chmod +x utils/log)
```bash
#!/usr/bin/env bash
echo "[$PROJECT] [$(date +%FT%T)] $1"
```

`utils/compose` (chmod +x utils/compose)
```bash
#!/usr/bin/env bash
docker-compose --project $PROJECT --volume ${PROJECT_ROOT}:/srv ${@}
```

**Comment by jdx** [comment]
I will say that explicitly calling `source myutils.sh` might actually be a good thing. If you do that, shellcheck will be able to follow it (and if you're writing bash you should be using shellcheck)
